master fails in IE11

**Describe the bug**

We cannot currently do a release of `master`, because tests fail in IE11.

Using `git bisect` and `npm ci && npm run test-cloud`, I have arrived at: 

```
668c18b1e83a0bb1e8748d8a1655c66d4110e70f is the first bad commit
commit 668c18b1e83a0bb1e8748d8a1655c66d4110e70f
Author: Phred <fearphage@gmail.com>
Date:   Sat Sep 28 12:00:53 2019 -0500

    prevent walking falsy values

:040000 040000 5a9a5660ea4231fe49e29a9daa03d19ae625aadf e29066aa996d84feba50f02b4ae65c6981efddea M	lib
:040000 040000 530e83603e88f11c98dfedeb76405a4961cf071b 2dc8721602614948c1506ee37a160633705030e6 M	test
~
```


**To Reproduce**

1. Check out `master`
1. `npm ci`
1. `npm run test-cloud`

```
1) restore-object
       throws on falsy input:
     AssertionError: [assert.exception] Expected { message: "Trying to restore object but received false" } but threw Error (Trying to undefined object but received false)
Error: Trying to undefined object but received false
      at walkObject (lib/sinon/util/core/walk-object.js:8)
      at restoreObject (lib/sinon/restore-object.js:12)
      at Anonymous function (test/restore-object-test.js:25)
      at captureException (node_modules/@sinonjs/referee/lib/capture-exception.js:5)
      at assert (node_modules/@sinonjs/referee/lib/assertions/exception.js:32)
      at assertion (node_modules/@sinonjs/referee/lib/define-assertion.js:51)
      at referee[type][name] (node_modules/@sinonjs/referee/lib/define-assertion.js:93)
      at Anonymous function (test/restore-object-test.js:23)
   at Runnable.prototype.run (Unknown script code:58624:7)
```

**Expected behavior**

No failed tests in browsers

--- a/lib/sinon/util/core/walk-object.js
+++ b/lib/sinon/util/core/walk-object.js
@@ -1,16 +1,13 @@
 "use strict";
 
-var valueToString = require("@sinonjs/commons").valueToString;
+var functionName = require("@sinonjs/commons").functionName;
 
 var getPropertyDescriptor = require("./get-property-descriptor");
 var walk = require("./walk");
 
-var reFunctionName = /^function\s*([^\s(]+)/;
-
 function walkObject(predicate, object, filter) {
     var called = false;
-    // IE11 doesn't support fn.name
-    var name = predicate.name || (valueToString(predicate).match(reFunctionName) || [null, ""])[1];
+    var name = functionName(predicate);
 
     if (!object) {
         throw new Error("Trying to " + name + " object but received " + String(object));

--- a/lib/sinon/util/core/walk-object.js
+++ b/lib/sinon/util/core/walk-object.js
@@ -1,13 +1,19 @@
 "use strict";
 
+var valueToString = require("@sinonjs/commons").valueToString;
+
 var getPropertyDescriptor = require("./get-property-descriptor");
 var walk = require("./walk");
 
+var reFunctionName = /^function\s*([^\s(]+)/;
+
 function walkObject(predicate, object, filter) {
     var called = false;
+    // IE11 doesn't support fn.name
+    var name = predicate.name || (valueToString(predicate).match(reFunctionName) || [null, ""])[1];
 
     if (!object) {
-        throw new Error("Trying to " + predicate.name + " object but received " + String(object));
+        throw new Error("Trying to " + name + " object but received " + String(object));
     }
 
     walk(object, function(prop, propOwner) {
@@ -31,7 +37,7 @@ function walkObject(predicate, object, filter) {
     });
 
     if (!called) {
-        throw new Error("Expected to " + predicate.name + " methods on object but found none");
+        throw new Error("Expected to " + name + " methods on object but found none");
     }
 
     return object;

--- a/test/util/core/walk-object-test.js
+++ b/test/util/core/walk-object-test.js
@@ -48,14 +48,14 @@ describe("util/core/walk-object", function() {
                 function() {
                     walkObject(anonymousFn, false);
                 },
-                { message: "Trying to  object but received false" }
+                { message: "Trying to undefined object but received false" }
             );
 
             assert.exception(
                 function() {
                     walkObject(anonymousFn, {});
                 },
-                { message: "Expected to  methods on object but found none" }
+                { message: "Expected to undefined methods on object but found none" }
             );
         });
     });

--- a/test/util/core/walk-object-test.js
+++ b/test/util/core/walk-object-test.js
@@ -17,7 +17,11 @@ describe("util/core/walk-object", function() {
         };
 
         before(function() {
-            if (typeof Object.defineProperty !== "function") {
+            if (
+                typeof Object.defineProperty !== "function" ||
+                typeof Object.getOwnPropertyDescriptor !== "function" ||
+                (Object.getOwnPropertyDescriptor(fnWithNoName, "name") || {}).configurable !== true
+            ) {
                 this.skip();
             }
 

--- a/test/util/core/walk-object-test.js
+++ b/test/util/core/walk-object-test.js
@@ -1,0 +1,62 @@
+"use strict";
+
+var referee = require("@sinonjs/referee");
+
+var walkObject = require("../../../lib/sinon/util/core/walk-object");
+
+var assert = referee.assert;
+
+describe("util/core/walk-object", function() {
+    // IE11
+    describe("without function.name support", function() {
+        function fnWithNoName() {
+            return;
+        }
+        var anonymousFn = function() {
+            return;
+        };
+
+        before(function() {
+            if (typeof Object.defineProperty !== "function") {
+                this.skip();
+            }
+
+            var descriptor = { value: undefined };
+
+            Object.defineProperty(anonymousFn, "name", descriptor);
+            Object.defineProperty(fnWithNoName, "name", descriptor);
+        });
+
+        it("should still identify functions in environments", function() {
+            assert.exception(
+                function() {
+                    walkObject(fnWithNoName, false);
+                },
+                { message: "Trying to fnWithNoName object but received false" }
+            );
+
+            assert.exception(
+                function() {
+                    walkObject(fnWithNoName, {});
+                },
+                { message: "Expected to fnWithNoName methods on object but found none" }
+            );
+        });
+
+        it("should work with anonymous functions", function() {
+            assert.exception(
+                function() {
+                    walkObject(anonymousFn, false);
+                },
+                { message: "Trying to  object but received false" }
+            );
+
+            assert.exception(
+                function() {
+                    walkObject(anonymousFn, {});
+                },
+                { message: "Expected to  methods on object but found none" }
+            );
+        });
+    });
+});

